backends-common/elasticsearch-v7/src/main/java/org/apache/james/backends/es/v7/IndexCreationFactory.java

```diff
@@ -33,65 +36,121 @@
+import com.fasterxml.jackson.databind.ObjectMapper;
 import com.github.fge.lambdas.Throwing;
 import com.google.common.base.Preconditions;
 import com.google.common.collect.ImmutableList;
 
 public class IndexCreationFactory {
 
-    public static class AliasSpecificationStep {
-        private final int nbShards;
-        private final int nbReplica;
-        private final int waitForActiveShards;
-        private final IndexName indexName;
-        private final ImmutableList.Builder<AliasName> aliases;
+    public static class IndexCreationCustomElement {
+        public static IndexCreationCustomElement EMPTY = from("{}");
 
-        AliasSpecificationStep(int nbShards, int nbReplica, int waitForActiveShards, IndexName indexName) {
-            this.nbShards = nbShards;
-            this.nbReplica = nbReplica;
-            this.waitForActiveShards = waitForActiveShards;
-            this.indexName = indexName;
-            this.aliases = ImmutableList.builder();
+        public static IndexCreationCustomElement from(String value) {
+            try {
+                new ObjectMapper().readTree(value);
+            } catch (JsonProcessingException e) {
+                throw new IllegalArgumentException("value must be a valid json");
+            }
+            return new IndexCreationCustomElement(value);
         }
 
-        public AliasSpecificationStep addAlias(AliasName aliasName) {
-            Preconditions.checkNotNull(aliasName);
-            this.aliases.add(aliasName);
-            return this;
-        }
+        private final String payload;
 
-        public ReactorElasticSearchClient createIndexAndAliases(ReactorElasticSearchClient client) {
-            return new IndexCreationPerformer(nbShards, nbReplica, waitForActiveShards, indexName, aliases.build()).createIndexAndAliases(client, Optional.empty());
+        IndexCreationCustomElement(String payload) {
+            this.payload = payload;
         }
 
-        public ReactorElasticSearchClient createIndexAndAliases(ReactorElasticSearchClient client, XContentBuilder mappingContent) {
-            return new IndexCreationPerformer(nbShards, nbReplica, waitForActiveShards, indexName, aliases.build()).createIndexAndAliases(client, Optional.of(mappingContent));
+        public String getPayload() {
+            return payload;
         }
     }
 
     static class IndexCreationPerformer {
+        public static class Builder {
+            private final int nbShards;
+            private final int nbReplica;
+            private final int waitForActiveShards;
+            private final IndexName indexName;
+            private final ImmutableList.Builder<AliasName> aliases;
+            private Optional<IndexCreationCustomElement> customAnalyzers;
+            private Optional<IndexCreationCustomElement> customTokenizers;
+
+            public Builder(int nbShards, int nbReplica, int waitForActiveShards, IndexName indexName) {
```

Comment:
I find your builder a bit strange TBH. It does not follow the usual syntax of other builders in the code base

Reply:
yes, some properties are immutable, and they will be inited by the constructor.

Reply:
Don't we do staged builds then for immutable properties? (well I think^^')

Reply:
Can you give me a code example? 
Create a new pojo object for immutable properties?

Reply:
@chibenwa wrote an article on medium a while ago: https://medium.com/linagora-engineering/next-level-java-8-staged-builders-602530f68b75

You can pretty much force the steps on your mandatory immutable values with it. You have numerous examples in the code. A small one for example is the builder in `StorageInformation` class :)

Reply:
Ah, I remember it, 
yesterday, 2 points make me didn't choose it. 
- `.addAlias` can recall multi-time. (append to list). I think it will be better if we needn't change anywhere already using it. (example: org.apache.james.mailbox.elasticsearch.v7.MailboxIndexCreationUtil)
- `customAnalyzers` and `customTokenizers` are optional, and we needn't declare each time

Reply:
for optional fields, like `customAnalyzers` and `customTokenizers`, you can mix staged builder and normal one. I think we have such builders in the code (like in the builder for `DeletedMessage` class, you have all params part of steps, except the last one, subject, that is optional)

I think for `addAlias` it is the same principle than the optional fields. Point is you can mix :)

Reply:
> customAnalyzers and customTokenizers are optional, and we needn't declare each time

Then just pass them to a Final Stage as a normal builder ;-)